File: Striver/Trees/Q23.java
package Striver.Trees;

public class Q23{
    // Lowest Common Ancestor of a Binary Tree

    public Node lowestCommonAncestor(Node root, Node p, Node q) {
        if(root == null || root == p || root == q) return root;
        
        Node left = lowestCommonAncestor(root.left, p, q);
        Node right = lowestCommonAncestor(root.right, p, q);
        
        if(left == null) return right;
        if(right == null) return left;

        return root;
    }

    public static void main(String[] args) {
        Node root = new Node(3);
        root.left = new Node(5);
        root.right = new Node(1);
        root.left.left = new Node(6);
        root.left.right = new Node(2);
        root.right.left = new Node(0);
        root.right.right = new Node(8);
        root.left.right.left = new Node(7);
        root.left.right.right = new Node(4);

        Q23 q = new Q23();
        Node ans = q.lowestCommonAncestor(root, root.left, root.right);
        System.out.println(ans.data);
    }
}

File: Striver/Trees/Q23.py
# lowest common ancestor in a binary tree

from Node import Node

def lca(root, n1, n2):
    if root is None or root.data == n1 or root.data == n2:
        return root
    
    left = lca(root.left, n1, n2)
    right = lca(root.right, n1, n2)

    if left is None:
        return right
    
    if right is None:
        return left
    
    return root

root = Node(10)
root.left = Node(20)
root.right = Node(30)
root.left.left = Node(40)
root.left.right = Node(50)

print(lca(root, 40, 50).data)